public/app/features/explore/ContentOutline/ContentOutlineContext.tsx

```diff
@@ -1,64 +1,163 @@
 import { uniqueId } from 'lodash';
-import React, { useState, useContext, createContext, ReactNode, useCallback } from 'react';
+import React, { useState, useContext, createContext, ReactNode, useCallback, useRef, useEffect } from 'react';
 
 import { ContentOutlineItemBaseProps } from './ContentOutlineItem';
 
 export interface ContentOutlineItemContextProps extends ContentOutlineItemBaseProps {
   id: string;
   ref: HTMLElement | null;
+  children?: ContentOutlineItemContextProps[];
 }
 
-type RegisterFunction = ({ title, icon, ref }: Omit<ContentOutlineItemContextProps, 'id'>) => string;
+type RegisterFunction = (outlineItem: Omit<ContentOutlineItemContextProps, 'id'>) => string;
 
 export interface ContentOutlineContextProps {
   outlineItems: ContentOutlineItemContextProps[];
   register: RegisterFunction;
   unregister: (id: string) => void;
+  updateOutlineItems: (newItems: ContentOutlineItemContextProps[]) => void;
+}
+
+interface ContentOutlineContextProviderProps {
+  children: ReactNode;
+  /**
+   * used to resort children of an outline item when the dependencies change
+   * e.g. when the order of query rows changes on drag and drop
+   */
+  refreshDependencies?: any[];
```

Comment:
can this be
```suggestion
  refreshDependencies?: unknown[];
```
instead?

Reply:
ah yes, unknown is much better for type safety